src/problems.jl

```diff
@@ -1,81 +1,41 @@
-mutable struct Solution{T<:Number}
-    primal::Array{T, 1}
-    dual::Array{T, 1}
-    status::Symbol
-    optval::T
-    has_dual::Bool
-end
-
-Solution(x::Array{T, 1}, status::Symbol, optval::T) where {T} =
-    Solution(x, T[], status, optval, false)
-Solution(x::Array{T, 1}, y::Array{T, 1}, status::Symbol, optval::T) where {T} =
-    Solution(x, y, status, optval, true)
-
-mutable struct Problem
+mutable struct Problem{T<:Real}
     head::Symbol
     objective::AbstractExpr
     constraints::Array{Constraint}
-    status::Symbol
-    optval::Float64OrNothing
-    model::Union{MathProgBase.AbstractConicModel, Nothing}
-    solution::Solution
+    status::MOI.TerminationStatusCode
+    model::Union{MOI.ModelLike, Nothing}
 
-    function Problem(head::Symbol, objective::AbstractExpr,
-                     model::Union{MathProgBase.AbstractConicModel, Nothing},
-                     constraints::Array=Constraint[])
+    function Problem{T}(head::Symbol, objective::AbstractExpr,
+                     constraints::Array=Constraint[]) where {T <: Real}
         if sign(objective)== Convex.ComplexSign()
-            error("Objective can not be a complex expression")
+            error("Objective cannot be a complex expression")
         else
-            return new(head, objective, constraints, Symbol("not yet solved"), nothing, model)
+            return new(head, objective, constraints, MOI.OPTIMIZE_NOT_CALLED, nothing)
         end
     end
 end
 
-# constructor if model is not specified
-function Problem(head::Symbol, objective::AbstractExpr, constraints::Array=Constraint[],
-                 solver::Union{MathProgBase.AbstractMathProgSolver, Nothing}=nothing)
-    model = solver !== nothing ? MathProgBase.ConicModel(solver) : solver
-    Problem(head, objective, model, constraints)
-end
-
-# If the problem constructed is of the form Ax=b where A is m x n
-# returns:
-# index: n
-# constr_size: m
-# var_to_ranges a dictionary mapping from variable id to (start_index, end_index)
-# where start_index and end_index are the start and end indexes of the variable in A
-function find_variable_ranges(constraints, id_to_variables)
-    index = 0
-    constr_size = 0
-    var_to_ranges = Dict{UInt64, Tuple{Int, Int}}()
-    for constraint in constraints
-        for i = 1:length(constraint.objs)
-            for (id, val) in constraint.objs[i]
-                if !haskey(var_to_ranges, id) && id != objectid(:constant)
-                    var = id_to_variables[id]
-                    if var.sign == ComplexSign()
-                        var_to_ranges[id] = (index + 1, index + 2*length(var))
-                        index += 2*length(var)
-                    else
-                        var_to_ranges[id] = (index + 1, index + length(var))
-                        index += length(var)
-                    end
-                end
-            end
-            constr_size += constraint.sizes[i]
+function Base.getproperty(p::Problem, s::Symbol)
```

Comment:
I'm open to feedback on this one. The reason I made `optval` lazy in this way is so that we don't need to query `MOI.ObjectiveValue()` during the `solve!` command (which would be eagerly populating `optval`). That way if the solver did not find a solution and hence may throw if you ask for its objective value, we don't error during `solve!` and you still get the status and termination info, along with whatever else the solver returned. This also matches how JuMP does it.

We could query to ask MOI if there is a primal feasible solution etc and only populate under those circumstances as well. It wasn't entirely clear to me exactly what conditions will guarantee `get(model, MOI.ObjectiveValue())` will not throw, however.

Reply:
Is this the only instance where you override `getproperty`? Why not deprecate `p.optval` in favor of `objective_value` that's implemented below?